Introduce Fragile interface. Implements #101

dmfs opened the pull request:
`Fragile` allows to defer checked Exceptions. Sometimes checked exceptions can't or shouldn't be thrown instantly. Common cases are background processes which deliver a result or iterators which don't allow throwing checked Exceptions in `next()`.

This commit adds a few basic implementations and one use case of `Fragile`.

In addition this also tweaks the implementations of `Ascii` and `Hex` a bit.

Note, a couple of unit tests are still missing.

dmfs commented:
@lemonboston I submitted this for discussion and comments not so much for merging. The question is, does the concept work out as expected with the given example implementations?

dmfs commented:
I'm wondering if it's possible and makes sense to make `Single` a subtype of `Fragile`, i.e. a Single is a special case of a 'Fragile` which doesn't throw. This way you could always pass or return a `Single` when a `Fragile` is expected. Maybe it makes sense to call it `FragileSingle` then instead of just `Fragile`. Just an idea.

dmfs commented:
I'm wondering if it's possible and makes sense to make `Single` a subtype of `Fragile`, i.e. a Single is a special case of a 'Fragile` which doesn't throw. This way you could always pass or return a `Single` when a `Fragile` is expected. Maybe it makes sense to call it `FragileSingle` then instead of just `Fragile`. Just an idea.

lemonboston commented:
I see that that could work, I guess it would simplify usage at some places by making `Single`s interchangeable with `Fragile`. On the other hand if we don't expect `Fragile` to be used often, I don't know if this is necessary. Even though I can't say any downside.

dmfs commented:
I would be a little concerned about semantics when having `interface Single extends Fragile`. The current definition of a `Single` is pretty clear: a set with just one element, a 1-tuple. This equals the mathematical definition. How would a `Fragile` have to be defined in order to preserve this definition of `Single`?

lemonboston commented:
I share this concern about semantical difference. It would feel safer to keep them separate. And when we need to convert between them, we have to do that explicitly, to stay conscious about what happening there and not allow it automatically.

dmfs commented:
@lemonboston I've removed most of the additional stuff and added a couple of unit tests. I'm still missing a few lines though. Two questions:
* Can we think of better names for the elementary `Fragile` implementations?
* Should we combine `SingleFragile` and `StableFragile` into one class?

lemonboston commented:
How about `Success` and `Failure` for `Single/StableFragile` (yes, I think it better as one class) and `ThrowingFragile`?

I also wondered a bit about this topic in general, how to represent 'result' types which are either a success or a failure. One thing is that I think using Exception may not fit well in general. For example if we catch an Exception somewhere but want to pass it further up where errors are handled, we need to again use an Exception (either keep it or wrap and ret-hrow). So we are stuck with representing failures with Exception semantics. Instead of having general error type so we can define our own when needed, with some error code, message, cause Exception, whatever.
And indeed I found now that in Scala they have two types for it: `Try` with the Exception and `Either` with the general success/error type.
http://blog.xebia.com/try-option-or-either/
(the TLDR section at the bottom is good)

This Kotlin library is inspired by Scala and it has it as well:
http://arrow-kt.io/docs/datatypes/try/

Some more links I found regarding Java:
https://github.com/lambdista/try
https://github.com/aol/cyclops/wiki/try-:-functional-exception-handling-for-java-8
https://stackoverflow.com/questions/26162407/is-there-an-equivalent-of-scalas-either-in-java-8

I am just mentioning these as something that may be interesting. This `Fragile` type is a bit different in this form.

lemonboston commented:
For some reason I don't see the `Encoded` class in the diff here.. I just wanted to comment whether you would like to add a secondary constructor with `CharSequence`.

dmfs commented:
I didn't include Encoded in the PR. I'll revisit this stuff and create a separate issue for that.

dmfs commented:
I'll look into this `Either` concept. Please create a story regarding this "SuccessOrFailure" type so we can move the discussion there.

lemonboston commented:
> I didn't include Encoded in the PR. I'll revisit this stuff and create a separate issue for that.

Sorry, I just hadn't had the latest version of the branch that's why I still saw it there.

Created the discussion ticket for SuccessOrFailure type: https://github.com/dmfs/jems/issues/150

Regarding the names of `Fragile`s here, how about just `Stable` and `Throwing` then?
In a similar way that we don't include the interface name for the base `Optional` types `Present` and `Absent` either, we could use shorter names here as well.
`Stable` would have the 2 ctors in this case.

dmfs commented:
I can't get used to `Stable`, how about `Safe`, `Solid`, `Firm`, `Sturdy` or `Strong`?
`Throwing` makes more sense.
On the other hand, when I think of a fragile object (like a porcelain vase) I'd say it can have two states, `Intact` and `Broken`. So how about these?

dmfs commented:
@lemonboston I've made a couple of changes and renamed the classes, please check it out.

dmfs commented:
@lemonboston I've added two more implementations of `Fragile` to demonstrate and test the concept. `Zipped` revealed a minor issue with the generic parameters of `Intact`.

lemonboston commented on src/main/java/org/dmfs/jems/fragile/composite/Zipped.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author Marten Gajda
+ */
+public final class Zipped<Left, Right, Result, E extends Throwable> implements Fragile<Result, E>
```
Can't you use generic type parameters for all 3 exceptions? Like (it compiles for me):
```java
public final class Zipped<Left, Right, Result, ResultException extends Throwable, LeftException extends ResultException, RightException extends ResultException> implements Fragile<Result, ResultException>
{
    private final BiFunction<Left, Right, Result> mZipFunction;
    private final Fragile<Left, LeftException> mLeft;
    private final Fragile<Right, RightException> mRight;


    public Zipped(Fragile<Left, LeftException> left, Fragile<Right, RightException> right, BiFunction<Left, Right, Result> zipFunction)
    {
        mZipFunction = zipFunction;
        mLeft = left;
        mRight = right;
    }


    @Override
    public Result value() throws ResultException
    {
        return mZipFunction.value(mLeft.value(), mRight.value());
    }
}
```

dmfs replied:
Ok, but what's the advantage? We don't really care about `LeftException` and `RightException` other than that `ResultException` has to be a super type of both of them. They both can be anonymous, we don't need a name for them and the unit test seems to confirm that the concept works.

lemonboston replied:
Right, I overlooked that it's not just `?` but `? extends E` so the super type thing is enforced this way as well.
What was the slight issue then what you mentioned?

dmfs replied:
`Intact` had a hard coded `Exception` as the generic `Throwable` parameter.

lemonboston replied:
Oh, I see, so that was a fix.

lemonboston commented on jems-testing/src/main/java/org/dmfs/jems/hamcrest/matchers/FragileMatcher.java:
```diff
@@ -0,0 +1,105 @@
+
+    public static <E extends Throwable> Matcher<Fragile<?, E>> isBroken(Class<E> exception)
+    {
+        return new FeatureMatcher<Fragile<?, E>, E>(instanceOf(exception), "broken Fragile throwing",
```
This class serves both as `Matcher` implementation and has factory for another `Matcher` created here as anonymous class.
How about making the class util class with the factory methods and having 2 inner class `Matcher`s? The `isIntact()` can also be simpler with `FeatureMatcher` probably.

dmfs replied:
With a `FeatureMatcher` I couldn't return `false` if the tested `Fragile` throws. I'm actually considering to replace `isBroken` with a regular implementation as well.

dmfs commented:
@lemonboston please check again. I've split `FragileMatcher` into two to make it more clear.
